<a href="https://colab.research.google.com/github/annisaeka123/2341720131_ML_2025/blob/main/JS13/TP_JS13.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Tugas Praktikum

Gunakan JST untuk klasifikasi angka tulisan tangan (MNIST).

Langkah:

- Load dataset MNIST dari Keras.

- Bangun model dengan 2 hidden layer.

- Latih model dan evaluasi akurasi.

In [3]:
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten
from tensorflow.keras.utils import to_categorical

# 1. Load dataset MNIST
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# Normalisasi data (0-255 → 0-1)
X_train = X_train / 255.0
X_test = X_test / 255.0

# One-hot encoding label
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

# 2. Bangun model JST
model = Sequential([
    Flatten(input_shape=(28, 28)),       # Ubah 28x28 → vektor 784
    Dense(128, activation='relu'),       # Hidden layer 1
    Dense(64, activation='relu'),        # Hidden layer 2
    Dense(10, activation='softmax')      # Output 10 kelas
])

# 3. Kompilasi model
model.compile(
    optimizer='adam',
    loss='categorical_crossentropy',
    metrics=['accuracy']
)

# 4. Latih model
model.fit(
    X_train, y_train,
    epochs=10,
    batch_size=32,
    validation_split=0.1,
    verbose=1
)

# 5. Evaluasi model
loss, acc = model.evaluate(X_test, y_test)
print(f"Akurasi pada data uji: {acc:.4f}")

11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step


/usr/local/lib/python3.12/dist-packages/keras/src/layers/reshaping/flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 1/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step - accuracy: 0.8633 - loss: 0.4558 - val_accuracy: 0.9650 - val_loss: 0.1217
Epoch 2/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 8s 5ms/step - accuracy: 0.9662 - loss: 0.1094 - val_accuracy: 0.9735 - val_loss: 0.0925
Epoch 3/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.9788 - loss: 0.0699 - val_accuracy: 0.9735 - val_loss: 0.0883
Epoch 4/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 8s 5ms/step - accuracy: 0.9836 - loss: 0.0515 - val_accuracy: 0.9752 - val_loss: 0.0850
Epoch 5/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.9881 - loss: 0.0384 - val_accuracy: 0.9753 - val_loss: 0.0855
Epoch 6/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 8s 5ms/step - accuracy: 0.9899 - loss: 0.0313 - val_accuracy: 0.9797 - val_loss: 0.0713
Epoch 7/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 8s 5ms/step - accuracy: 0.9926 - loss: 0.0231 - val_accuracy: 0.9790 - val_loss: 0.0750
Epoch 8/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.9926 - loss: 0.0206 - 

Hidden Layer Lebih Besar (256 → 128)

In [4]:
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten
from tensorflow.keras.utils import to_categorical

# 1. Load dataset MNIST
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# Normalisasi data (0-255 → 0-1)
X_train = X_train / 255.0
X_test = X_test / 255.0

# One-hot encoding label
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

# 2. Bangun model JST
model = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(256, activation='relu'),
    Dense(128, activation='relu'),
    Dense(10, activation='softmax')
])

# 3. Kompilasi model
model.compile(
    optimizer='adam',
    loss='categorical_crossentropy',
    metrics=['accuracy']
)

# 4. Latih model
model.fit(
    X_train, y_train,
    epochs=10,
    batch_size=32,
    validation_split=0.1,
    verbose=1
)

# 5. Evaluasi model
loss, acc = model.evaluate(X_test, y_test)
print(f"Akurasi pada data uji: {acc:.4f}")

Epoch 1/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 12s 6ms/step - accuracy: 0.8894 - loss: 0.3793 - val_accuracy: 0.9655 - val_loss: 0.1126
Epoch 2/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 10s 6ms/step - accuracy: 0.9711 - loss: 0.0919 - val_accuracy: 0.9737 - val_loss: 0.0867
Epoch 3/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 9s 6ms/step - accuracy: 0.9816 - loss: 0.0563 - val_accuracy: 0.9755 - val_loss: 0.0804
Epoch 4/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 11s 6ms/step - accuracy: 0.9865 - loss: 0.0412 - val_accuracy: 0.9797 - val_loss: 0.0796
Epoch 5/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 10s 6ms/step - accuracy: 0.9898 - loss: 0.0314 - val_accuracy: 0.9780 - val_loss: 0.0818
Epoch 6/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 11s 6ms/step - accuracy: 0.9911 - loss: 0.0264 - val_accuracy: 0.9730 - val_loss: 0.1071
Epoch 7/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 11s 6ms/step - accuracy: 0.9927 - loss: 0.0223 - val_accuracy: 0.9778 - val_loss: 0.1014
Epoch 8/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 10s 6ms/step - accuracy: 0.9934 - loss: 0.

Tambah 1 Hidden Layer Lagi (3 hidden layers)

In [5]:
# 1. Load dataset MNIST
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# Normalisasi data (0-255 → 0-1)
X_train = X_train / 255.0
X_test = X_test / 255.0

# One-hot encoding label
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

# 2. Bangun model JST
model = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(256, activation='relu'),
    Dense(128, activation='relu'),
    Dense(64, activation='relu'),
    Dense(10, activation='softmax')
])

# 3. Kompilasi model
model.compile(
    optimizer='adam',
    loss='categorical_crossentropy',
    metrics=['accuracy']
)

# 4. Latih model
model.fit(
    X_train, y_train,
    epochs=10,
    batch_size=32,
    validation_split=0.1,
    verbose=1
)

# 5. Evaluasi model
loss, acc = model.evaluate(X_test, y_test)
print(f"Akurasi pada data uji: {acc:.4f}")

Epoch 1/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 12s 6ms/step - accuracy: 0.8788 - loss: 0.4007 - val_accuracy: 0.9687 - val_loss: 0.1027
Epoch 2/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 11s 6ms/step - accuracy: 0.9706 - loss: 0.0998 - val_accuracy: 0.9700 - val_loss: 0.0944
Epoch 3/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 11s 6ms/step - accuracy: 0.9803 - loss: 0.0630 - val_accuracy: 0.9777 - val_loss: 0.0726
Epoch 4/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 9s 5ms/step - accuracy: 0.9854 - loss: 0.0449 - val_accuracy: 0.9797 - val_loss: 0.0717
Epoch 5/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 11s 6ms/step - accuracy: 0.9885 - loss: 0.0371 - val_accuracy: 0.9810 - val_loss: 0.0740
Epoch 6/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 10s 6ms/step - accuracy: 0.9906 - loss: 0.0306 - val_accuracy: 0.9810 - val_loss: 0.0747
Epoch 7/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 10s 6ms/step - accuracy: 0.9917 - loss: 0.0264 - val_accuracy: 0.9807 - val_loss: 0.0746
Epoch 8/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 10s 6ms/step - accuracy: 0.9934 - loss: 0.

Perbandingan Akurasi dan Waktu Pelatihan

Eksperimen dilakukan dengan tiga konfigurasi arsitektur jaringan saraf tiruan (JST) berbeda pada dataset MNIST. Tujuannya adalah membandingkan pengaruh jumlah neuron dan jumlah hidden layer terhadap akurasi serta waktu pelatihan.

Tabel Perbandingan


| Model  | Konfigurasi Hidden Layer | Total Waktu Training (10 epoch) | Akurasi Test |
| ------ | ------------------------ | ------------------------------- | ------------ |
| **M1** | 128 → 64                 | ~80–85 detik                    | **0.9786**   |
| **M2** | 256 → 128                | ~100–110 detik                  | **0.9799**   |
| **M3** | 256 → 128 → 64           | ~100–120 detik                  | **0.9709**   |


a. Konfigurasi Dasar (128 → 64)

Model pertama menggunakan dua hidden layer dengan jumlah neuron 128 dan 64.

- Total waktu pelatihan: ± 80–85 detik

- Akurasi data uji: 0.9786

Model ini memiliki waktu pelatihan tercepat dan akurasinya cukup tinggi, sehingga dapat dianggap sebagai konfigurasi yang efisien.

b. Hidden Layer Lebih Besar (256 → 128)

Model kedua meningkatkan jumlah neuron pada kedua hidden layer menjadi 256 dan 128.

- Total waktu pelatihan: ± 100–110 detik

- Akurasi data uji: 0.9799

Model ini menghasilkan akurasi tertinggi dibandingkan konfigurasi lainnya. Waktu pelatihan sedikit lebih lama, tetapi peningkatan akurasi cukup signifikan.

c. Penambahan Hidden Layer Baru (256 → 128 → 64)

Model ketiga menambahkan satu hidden layer tambahan (64 neuron).

- Total waktu pelatihan: ± 100–120 detik

- Akurasi data uji: 0.9709

Meskipun memiliki arsitektur paling kompleks, model ini justru menunjukkan penurunan akurasi pada data uji. Hal ini mengindikasikan terjadinya overfitting, di mana model terlalu menyesuaikan diri dengan data pelatihan dan kehilangan generalisasi.

Kesimpulan

Dari hasil pengujian, dapat disimpulkan bahwa:

- Penambahan jumlah neuron meningkatkan akurasi tanpa memberikan lonjakan waktu pelatihan yang signifikan.

- Penambahan hidden layer berlebih tidak selalu memberikan performa lebih baik, bahkan dapat menyebabkan overfitting sehingga akurasi menurun.

- Hidden Layer Lebih Besar (256 → 128) merupakan konfigurasi paling optimal pada eksperimen ini, dengan keseimbangan terbaik antara akurasi dan waktu pelatihan.

Aktivasi Sigmoid (dibandingkan ReLU)

In [6]:
# 1. Load dataset MNIST
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# Normalisasi data (0-255 → 0-1)
X_train = X_train / 255.0
X_test = X_test / 255.0

# One-hot encoding label
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

# 2. Bangun model JST
model = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(256, activation='sigmoid'),
    Dense(128, activation='sigmoid'),
    Dense(10, activation='softmax')
])

# 3. Kompilasi model
model.compile(
    optimizer='adam',
    loss='categorical_crossentropy',
    metrics=['accuracy']
)

# 4. Latih model
model.fit(
    X_train, y_train,
    epochs=10,
    batch_size=32,
    validation_split=0.1,
    verbose=1
)

# 5. Evaluasi model
loss, acc = model.evaluate(X_test, y_test)
print(f"Akurasi pada data uji: {acc:.4f}")

Epoch 1/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 11s 6ms/step - accuracy: 0.8127 - loss: 0.7213 - val_accuracy: 0.9557 - val_loss: 0.1625
Epoch 2/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 10s 6ms/step - accuracy: 0.9465 - loss: 0.1826 - val_accuracy: 0.9690 - val_loss: 0.1126
Epoch 3/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 9s 5ms/step - accuracy: 0.9646 - loss: 0.1193 - val_accuracy: 0.9732 - val_loss: 0.0886
Epoch 4/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 10s 6ms/step - accuracy: 0.9772 - loss: 0.0790 - val_accuracy: 0.9755 - val_loss: 0.0825
Epoch 5/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 10s 6ms/step - accuracy: 0.9824 - loss: 0.0599 - val_accuracy: 0.9765 - val_loss: 0.0775
Epoch 6/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 10s 6ms/step - accuracy: 0.9863 - loss: 0.0449 - val_accuracy: 0.9797 - val_loss: 0.0689
Epoch 7/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 10s 6ms/step - accuracy: 0.9909 - loss: 0.0322 - val_accuracy: 0.9798 - val_loss: 0.0683
Epoch 8/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 10s 6ms/step - accuracy: 0.9925 - loss: 0.

Perbandingan Fungsi Aktivasi: Sigmoid vs ReLU

a. Hasil Pelatihan dengan Aktivasi ReLU

- Akurasi pelatihan: 0.9943

- Akurasi validasi: ± 0.9800

- Akurasi data uji: 0.9786

- Waktu pelatihan per epoch: ± 7–9 detik

Model dengan ReLU sedikit lebih cepat dalam pelatihan dan stabil dalam menangani gradien. ReLU mencegah masalah vanishing gradient, sehingga pelatihan lebih efisien.

b. Hasil Pelatihan dengan Aktivasi Sigmoid

- Berdasarkan hasil training:

- Akurasi pelatihan mencapai: 0.9961

- Akurasi validasi terbaik: ± 0.9812

- Akurasi data uji: 0.9790

- Waktu pelatihan per epoch: ± 10–11 detik

Model dengan Sigmoid menunjukkan performa yang baik, namun waktu pelatihannya lebih lama karena Sigmoid memiliki operasi eksponensial dan rentan mengalami vanishing gradient, terutama pada lapisan dalam.

Ringkasan Perbandingan ReLU vs Sigmoid

| Aspek                    | Sigmoid                            | ReLU                          |
| ------------------------ | ---------------------------------- | ----------------------------- |
| Akurasi Uji              | **0.9790**                         | 0.9786                        |
| Kecepatan Pelatihan      | Lebih lambat (10–11 detik/epoch)   | Lebih cepat (7–9 detik/epoch) |
| Stabilitas Gradien       | Rentan vanishing gradient          | Stabil dan efisien            |
| Kompleksitas Perhitungan | Lebih tinggi (fungsi eksponensial) | Lebih rendah (max(0, x))      |
| Cocok Untuk              | Model kecil, output probabilistik  | Deep learning modern          |


Kesimpulan

- Sigmoid menghasilkan akurasi sedikit lebih tinggi daripada ReLU pada eksperimen ini, namun perbedaannya sangat kecil.

- ReLU lebih cepat dilatih, karena operasinya sederhana dan tidak mengalami saturasi negatif.

- Untuk model yang lebih besar atau lebih dalam, ReLU tetap lebih direkomendasikan karena menghindari vanishing gradient.

- Pada kasus dataset MNIST dan arsitektur sederhana, kedua aktivasi memberikan performa hampir sama, sehingga pemilihan bergantung pada kecepatan pelatihan.